### PR TITLE
Remove atomic emulation

### PR DESCRIPTION
**Pull request by ZzEeKkAa**
Since DPCPP now supports atomics, we don't need to emulate them by ourselves.

- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

Fixes: #760, #650, #868, #1068

### CODE REVIEW

File: numba_dpex/ocl/oclimpl.py [hunk 1]
```diff
@@ -286,20 +229,15 @@ def native_atomic_add(context, builder, sig, args):
...
-            device_type == dpctl.device_type.gpu
-            and config.NATIVE_FP_ATOMICS == 1
-        ):
-            return native_atomic_add(context, builder, sig, args)
-        else:
-            # Currently, DPCPP only supports native floating point
-            # atomics for GPUs.
-            return atomic_add(context, builder, sig, args, "add")
-    elif dtype == types.int32 or dtype == types.int64:
+    # TODO: do we need this check, or should we just use native_atomic_add for everything?
```
Comment (diptorupd) [inline numba_dpex/ocl/oclimpl.py:234]: The check should check if atomic operations are supported for the type, and it should be based on what SYCL supports: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:atomic-references

Comment (ZzEeKkAa) [reply numba_dpex/ocl/oclimpl.py:234]: Does something like this work:
```python
def support_atomic(dtype: types.Type) -> bool:
    has_atomic64_support = True # TODO: get from dpctl

    # This check should be the same as described in sycl documentation:
    # https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:atomic-references
    return (dtype == types.int32
    or dtype == types.uint32
    or dtype == types.float32
    or (dtype == types.int64
    or dtype == types.uint64
    or dtype == types.float64)
    and has_atomic64_support)
```

File: numba_dpex/ocl/oclimpl.py [hunk 2]
```diff
@@ -282,27 +225,33 @@ def native_atomic_add(context, builder, sig, args):
...
-            device_type == dpctl.device_type.gpu
-            and config.NATIVE_FP_ATOMICS == 1
-        ):
-            return native_atomic_add(context, builder, sig, args)
-        else:
-            # Currently, DPCPP only supports native floating point
-            # atomics for GPUs.
-            return atomic_add(context, builder, sig, args, "add")
-    elif dtype == types.int32 or dtype == types.int64:
+    dtype, device = sig.args[0].dtype, sig.args[0].device
```
Comment (ZzEeKkAa) [inline numba_dpex/ocl/oclimpl.py:248]: @diptorupd do you know better way to get device here?

Comment (ZzEeKkAa) [reply numba_dpex/ocl/oclimpl.py]: I've figured it out

**Comment by ZzEeKkAa** [comment]
This PR should unblock these checks:
https://github.com/IntelPython/dpbench/pull/289

File: numba_dpex/examples/kernel/atomic_op.py [hunk 3]
```diff
@@ -10,12 +10,14 @@
 @ndpx.kernel
 def atomic_reduction(a):
     idx = ndpx.get_global_id(0)
```
Comment (diptorupd) [inline numba_dpex/examples/kernel/atomic_op.py:12]: Please add some docstring explaining what the example does. It should be pointed out to users the memory order (relaxed) and the scope (device) of the atomic operation.

File: numba_dpex/ocl/oclimpl.py [hunk 4]
```diff
@@ -282,27 +225,29 @@ def native_atomic_add(context, builder, sig, args):
     return builder.call(fn, fn_args)
 
 
+def support_atomic(dtype: types.Type) -> bool:
+    # This check should be the same as described in sycl documentation:
+    # https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:atomic-references
+    # If atomic is not supported, it will be emulated by the sycl compiler.
+    return (
```
Comment (diptorupd) [inline numba_dpex/ocl/oclimpl.py:232]: Fine for now. My follow up changes will make it possible to use `dpctl.SyclDevice.has_aspect_atomic64` to refine the check.

**Review by diptorupd — CHANGES_REQUESTED** [review]
The `run_test.sh` script in the `conda-recipe` directory should be updated as well.

The documentation page
https://github.com/IntelPython/numba-dpex/blob/main/docs/source/user_guide/kernel_programming/atomic-operations.rst should be updated as well.

Please test #1068, #774, #711 as well with this PR.

**Comment by diptorupd** [comment]
@ZzEeKkAa in a subsequent PR can you push changes to remove the dependency on spirv-tools?

**Comment by ZzEeKkAa** [comment]
#650 tested:
```cmd
> set NUMBA_DPPY_DEBUG=1
> dpbench -b gpairs -i numba_dpex_k run --no-skip-expected-failures

================ Benchmark GPairs (gpairs) ========================

================ implementation numba_dpex_k ========================
implementation: numba_dpex_k
framework: numba_dpex
framework version: 0.21.1+61.g7d9bcca0.dirty
input size: 8512
setup time: 219.6217ms (219621700 ns)
warmup time: 962.5928ms (962592800 ns)
teardown time: 101.4μs (101400 ns)
max execution times: 4.226ms (4226000 ns)
min execution times: 2.7134ms (2713400 ns)
median execution times: 3.15995ms (3159950 ns)
repeats: 10
preset: S
validated: Success
```

**Comment by ZzEeKkAa** [comment]
This change is ready to go. I've tested all the mentioned issues. 

These issues simply does not make sense now: #760, #868, #711 does not work, see my latest comment there.

**Comment by ZzEeKkAa** [comment]
> @ZzEeKkAa in a subsequent PR can you push changes to remove the dependency on spirv-tools?

Did it here in a separate commit.

**Pull request merged**